src/core/server/http/router/response_error.ts

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * Additional meta-data to enhance error output or provide error details.
+ * @public
+ */
+export interface ResponseErrorMeta {
+  data?: Record<string, any>;
+  errorCode?: string;
+  docLink?: string;
+}
+
+/**
+ * Kibana ResponseError object to store error details
+ * @public
+ */
+export class ResponseError extends Error {
+  constructor(error: Error | string, public readonly meta?: ResponseErrorMeta) {
+    super(typeof error === 'string' ? error : error.message);
+    Object.setPrototypeOf(this, ResponseError.prototype);
+  }
+}
+
+/**
+ * Creates Kibana ResponseError instance
+ * @param error - Error object or message to wrap
+ * @param meta - additional meta-data to pass to the client
+ * @public
+ */
+export const createResponseError = (error: Error | string, meta?: ResponseErrorMeta) =>
```

Comment:
because it is a stateless factory we can export is a helper from kibana root.
Usage:
```js
import { createResponseError } from 'kibana/server';
```

Reply:
> Seems like this would be easier to use and more discoverable if it was exposed to the response handlers

could you elaborate on this?

>  Alternatively, ResponseFactory could call this under the hood in the error methods.

hm. I thought about this and I see several options:
- we could extend a type definition of Error type to be able to declare additional properties. Btw that's the only reason why we introduced `ResponseError` type.
- pass additional data as an extra parameter to `responseFactory` error methods, which doesn't comply with other methods having a signature `(data to send, response options)`. And if an error and additional data are coupled, shouldn't we pass them around as one entity?
- pass additional tools as a separate argument:
```diff
export type RequestHandler<P extends ObjectType, Q extends ObjectType, B extends ObjectType> = (
  request: KibanaRequest<TypeOf<P>, TypeOf<Q>, TypeOf<B>>,
  response: ResponseFactory,
+ toolkit: { createError, ... }
) => KibanaResponse<any> | Promise<KibanaResponse<any>>;
```

Reply:
also I realised that the current solution forces to pass an error object, which is not always convenient. probably we should make it optional
```diff
- response.notFound(new Error('Cannot find requested'))
+ response.notFound()
```

Reply:
I like the first option, but may we could make it easier to consume by allowing the `error` argument to be either an Error or an object for configuring a ResponseError:

```diff
unauthorized: (
- error: ResponseError 
+ error: string | Error | { error?: string | Error, data?: ResponseErrorData }
  options: HttpResponseOptions = {}
)
```

Reply:
I missed this in the RFC (which might have been a better place to discuss it), but why do we want ResponseError to extend Error? I'm not sure how we would use the stack traces we get from Error to help debug or print out server logs.

In my head HTTP Error codes (except maybe a 500) aren't really errors or exceptions the same way "TypeError: Cannot read property 'foo' of undefined" is a sign that something is broken in your code. So these errors are just a value that your return to a client to explain why you couldn't do what they wanted you to do but all your business logic is working 100%.

Some API's for example return a JSON object with the error message, documentation links and other metadata. You could stringify this and wrap it in an error objects message, but it feels unnecessary. So I don't really see that much benefit in enforcing that non 2xx response codes should only take special error objects.